virtme/guest/virtme-sshd-script

```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+#
+# Initialize ssh server for remote connections (option `--server ssh`)
+
+HOME=$(getent passwd "${virtme_user}" | cut -d: -f6)
+cat ${HOME}/.ssh/id_rsa.pub >> ${HOME}/.ssh/authorized_keys
+chown ${virtme_user} ${HOME}/.ssh/authorized_keys
+mkdir -p /run/sshd
+rm -f /var/run/nologin
+/usr/sbin/sshd -h ~/.ssh/id_rsa
```

Comment:
Do you need `-h`? Is this script not going to be executed as root? If yes, then no need to specify `-h`, no?

If you need `-h`:
- can you use `~` instead of `${HOME}` here?
- also this RSA key might not be available if another algorithm is used

Reply:
We need to pass `-h`, otherwise sshd doesn't start and it complains about:
```
sshd: no hostkeys available -- exiting
```

I'll change the script to use `$HOME` (and also a bit smarter at finding the first private key available).

Reply:
> We need to pass `-h`, otherwise sshd doesn't start and it complains about:
> 
> ```
> sshd: no hostkeys available -- exiting
> ```

Strange, same here. I guess it is because it cannot use the keys that have been generated on the host. I can reproduce the issue on my side:

```
root@virtme-ng:/# /usr/sbin/sshd -dD
debug1: sshd version OpenSSH_9.9, OpenSSL 3.4.0 22 Oct 2024
debug1: Unable to load host key: /etc/ssh/ssh_host_rsa_key
debug1: Unable to load host key: /etc/ssh/ssh_host_ecdsa_key
debug1: Unable to load host key: /etc/ssh/ssh_host_ed25519_key
sshd: no hostkeys available -- exiting.
```

Here is a simple solution: replacing the hosts keys by new ones:

```
root@virtme-ng:/# rm /etc/ssh/ssh_host_*_key*
root@virtme-ng:/# /usr/bin/ssh-keygen -A
ssh-keygen: generating new host keys: RSA ECDSA ED25519 
root@virtme-ng:/# /usr/sbin/sshd -dD
debug1: sshd version OpenSSH_9.9, OpenSSL 3.4.0 22 Oct 2024
debug1: private host key #0: ssh-rsa SHA256:T4r1s6H3SUHnrr5ZNKRmIhoyoog8sTKbmia1W5Fw9PU
debug1: private host key #1: ecdsa-sha2-nistp256 SHA256:e0n9z3ot4KJZlSMQXV+EN0ZitMp5NhRmLpX+wx7q3po
debug1: private host key #2: ssh-ed25519 SHA256:++wwB2hFHUh8vBXKZ4S1DH17Yf0sAK6LeagcVIuHPVU
debug1: rexec_argv[1]='-dD'
debug1: Set /proc/self/oom_score_adj from 0 to -1000
debug1: Bind to port 22 on 0.0.0.0.
Server listening on 0.0.0.0 port 22.
debug1: Bind to port 22 on ::.
Server listening on :: port 22.
```

I guess instead of removing the keys, `ssh-keygen -A` should be used with `-f`, just in case `vng --rw` (or something else) is used :)

```
root@virtme-ng:/# TMP=$(mktemp -d)
root@virtme-ng:/# mkdir -p "${TMP}/etc/ssh"
root@virtme-ng:/# ssh-keygen -A -f "${TMP}"
ssh-keygen: generating new host keys: RSA ECDSA ED25519 
root@virtme-ng:/# ARGS=(); for i in "${TMP}"/etc/ssh/ssh_host_*_key; do ARGS+=(-h ${i}); done
root@virtme-ng:/# /usr/sbin/sshd "${ARGS[@]}"
```

> I'll change the script to use `$HOME` (and also a bit smarter at finding the first private key available).

Thanks!

Reply:
I forgot to add: on my side, I could not use `/usr/sbin/sshd -h ~/.ssh/id_XXX`, because my keys have a password.

Reply:
> > We need to pass `-h`, otherwise sshd doesn't start and it complains about:
> > ```
> > sshd: no hostkeys available -- exiting
> > ```
> 
> Strange, same here. I guess it is because it cannot use the keys that have been generated on the host. I can reproduce the issue on my side:

In my box it doesn't work, because as regular user I don't have access to my host's keys:
```
$ cat /etc/ssh/ssh_host_rsa_key
cat: /etc/ssh/ssh_host_rsa_key: Permission denied
```
 
> Here is a simple solution: replacing the hosts keys by new ones:

Yes, I was thinking about that, but the boot time will be slower, maybe we can try to use the user's keys if they exist and they're accessible, otherwise generate them on the fly.

Reply:
> > > We need to pass `-h`, otherwise sshd doesn't start and it complains about:
> > > ```
> > > sshd: no hostkeys available -- exiting
> > > ```
> > 
> > 
> > Strange, same here. I guess it is because it cannot use the keys that have been generated on the host. I can reproduce the issue on my side:
> 
> In my box it doesn't work, because as regular user I don't have access to my host's keys:
> 
> ```
> $ cat /etc/ssh/ssh_host_rsa_key
> cat: /etc/ssh/ssh_host_rsa_key: Permission denied
> ```

Yes, but the init script will be executed as root, no? (also needed to listen on port 22 by default)

> > Here is a simple solution: replacing the hosts keys by new ones:
> 
> Yes, I was thinking about that, but the boot time will be slower, maybe we can try to use the user's keys if they exist and they're accessible, otherwise generate them on the fly.

Loading the user's keys seems fragile: Which one to pick? What if there is a passphrase?

Generating host keys at boot time doesn't seem to be an issue, at least on my side:

``` shell
$ time sudo vng -r --user root -- 'mkdir -p /tmp/etc/ssh; ssh-keygen -A -f /tmp; /usr/sbin/sshd -h /tmp/etc/ssh/ssh_host_ed25519_key'
ssh-keygen: generating new host keys: RSA ECDSA ED25519 
sudo vng -r --user root --   0.00s user 0.01s system 0% cpu 2.384 total
$ time sudo vng -r --user root -- '/usr/sbin/sshd -h /root/.ssh/id_ed25519'
sudo vng -r --user root -- '/usr/sbin/sshd -h /root/.ssh/id_ed25519'  0.00s user 0.01s system 0% cpu 2.158 total
```

Also, to me, this SSH option is heavier than the Console one: it is normal if it takes longer, and it is more design to CI workflows, where 0.1 seconds more to be ready will not change much, no?

Reply:
> > > > We need to pass `-h`, otherwise sshd doesn't start and it complains about:
> > > > ```
> > > > sshd: no hostkeys available -- exiting
> > > > ```
> > > 
> > > 
> > > Strange, same here. I guess it is because it cannot use the keys that have been generated on the host. I can reproduce the issue on my side:
> > 
> > 
> > In my box it doesn't work, because as regular user I don't have access to my host's keys:
> > ```
> > $ cat /etc/ssh/ssh_host_rsa_key
> > cat: /etc/ssh/ssh_host_rsa_key: Permission denied
> > ```
> 
> Yes, but the init script will be executed as root, no? (also needed to listen on port 22 by default)

It is root in the guest, but on the host virtiofsd runs as regular user, so in practice we can't access the file, even if we're root in the guest. And that's good, otherwise we'd provide a way to bypass file permissions, which won't be really nice. :)

> 
> > > Here is a simple solution: replacing the hosts keys by new ones:
> > 
> > 
> > Yes, I was thinking about that, but the boot time will be slower, maybe we can try to use the user's keys if they exist and they're accessible, otherwise generate them on the fly.
> 
> Loading the user's keys seems fragile: Which one to pick? What if there is a passphrase?
> 
> Generating host keys at boot time doesn't seem to be an issue, at least on my side:

Yeah, the passphrase is a good point and generating them on-the-fly is probably more reliable. However, the annoying part about changing the keys every time is that ssh will complain about seeing different keys in `~/.known_hosts`.

Maybe we could pre-generate the guest keys and store them in `~/.cache/virtme-ng/.ssh/`? Then we pass the keys to the guest and we'll have safe ssh keys to use that we can access and we won't have known_hosts issues.

> Also, to me, this SSH option is heavier than the Console one: it is normal if it takes longer, and it is more design to CI workflows, where 0.1 seconds more to be ready will not change much, no?

Yes, ssh is certainly heavier than console, however some users have  workflows that involve transferring data to/from the VM via scp, so having an easy way to enable ssh access in the guest would be highly convenient for them and they could easily migrate their CI to use vng.

Reply:
> > > > > We need to pass `-h`, otherwise sshd doesn't start and it complains about:
> > > > > ```
> > > > > sshd: no hostkeys available -- exiting
> > > > > ```
> > > > 
> > > > 
> > > > Strange, same here. I guess it is because it cannot use the keys that have been generated on the host. I can reproduce the issue on my side:
> > > 
> > > 
> > > In my box it doesn't work, because as regular user I don't have access to my host's keys:
> > > ```
> > > $ cat /etc/ssh/ssh_host_rsa_key
> > > cat: /etc/ssh/ssh_host_rsa_key: Permission denied
> > > ```
> > 
> > 
> > Yes, but the init script will be executed as root, no? (also needed to listen on port 22 by default)
> 
> It is root in the guest, but on the host virtiofsd runs as regular user, so in practice we can't access the file, even if we're root in the guest. And that's good, otherwise we'd provide a way to bypass file permissions, which won't be really nice. :)

OK I see, in my case I launch vng from a docker container, with the root user. So both the host and the guest are root :)

Anyway, from the VM, best not to use the same host keys as the ones of the host, even if we are root and have access to them.

> > > > Here is a simple solution: replacing the hosts keys by new ones:
> > > 
> > > 
> > > Yes, I was thinking about that, but the boot time will be slower, maybe we can try to use the user's keys if they exist and they're accessible, otherwise generate them on the fly.
> > 
> > 
> > Loading the user's keys seems fragile: Which one to pick? What if there is a passphrase?
> > Generating host keys at boot time doesn't seem to be an issue, at least on my side:
> 
> Yeah, the passphrase is a good point and generating them on-the-fly is probably more reliable. However, the annoying part about changing the keys every time is that ssh will complain about seeing different keys in `~/.known_hosts`.
> 
> Maybe we could pre-generate the guest keys and store them in `~/.cache/virtme-ng/.ssh/`? Then we pass the keys to the guest and we'll have safe ssh keys to use that we can access and we won't have known_hosts issues.

Good idea!

> > Also, to me, this SSH option is heavier than the Console one: it is normal if it takes longer, and it is more design to CI workflows, where 0.1 seconds more to be ready will not change much, no?
> 
> Yes, ssh is certainly heavier than console, however some users have workflows that involve transferring data to/from the VM via scp, so having an easy way to enable ssh access in the guest would be highly convenient for them and they could easily migrate their CI to use vng.

I understand. So more an option to be compatible with older workflows, but not recommended for new ones. (Except if there is really a need for a remote host to communicate with the VM).